### ISSUE TITLE
Adam(fused=True) unexpected behaviour (massively decreased accuracy)

### ISSUE THREAD

**Issue by AlexSheldrick** [pytorch/pytorch#88258]
### 🐛 Describe the bug

I've tried the new Adam(fused=True) implementation and noticed a massive performance decrease in my project while using AMP. Just changing the flag from True to False reduces my loss by over 200% (in a mipnerf variant implemented in PyTorch-lightning). I wouldn't expect the fused kernel to perform drastically worse so it could be a bug in the implementation // interaction with AMP. Can share my repo if it helps with diagnosis.

### Versions

Collecting environment information...
PyTorch version: 1.13.0
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.9.12 (main, Jun  1 2022, 11:38:51)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.15.0-43-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2070
Nvidia driver version: 515.65.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.4
[pip3] pytorch-lightning==1.7.7
[pip3] torch==1.13.0
[pip3] torch-tb-profiler==0.4.0
[pip3] torchaudio==0.13.0
[pip3] torchmetrics==0.10.0
[pip3] torchvision==0.14.0
[conda] blas                      1.0                         mkl
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py39h7f8727e_0
[conda] mkl_fft                   1.3.1            py39hd3c417c_0
[conda] mkl_random                1.2.2            py39h51133e4_0
[conda] numpy                     1.23.4                   pypi_0    pypi
[conda] numpy-base                1.23.3           py39h31eccc5_0
[conda] pytorch                   1.13.0          py3.9_cuda11.7_cudnn8.5.0_0    pytorch
[conda] pytorch-cuda              11.7                 h67b0de4_0    pytorch
[conda] pytorch-lightning         1.7.7                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch                     1.12.1+cu113             pypi_0    pypi
[conda] torch-tb-profiler         0.4.0                    pypi_0    pypi
[conda] torchaudio                0.13.0               py39_cu117    pytorch
[conda] torchmetrics              0.10.0                   pypi_0    pypi
[conda] torchvision               0.13.1+cu113             pypi_0    pypi

cc @ezyang @gchanan @zou3519 @vincentqb @jbschlosser @albanD @ngimel @mcarilli @ptrblck @leslie-fang-intel @jgong5

**Comment by soumith** [comment]
sharing a reproduction would be really useful.

> Just changing the flag from True to False reduces my loss by over 200%

This makes me think that when you say "performance", you mean accuracy and not how fast the code is running?

**Comment by AlexSheldrick** [comment]
> This makes me think that when you say "performance", you mean accuracy and not how fast the code is running?

Yes, that's exactly right, I found my model converged to a much higher loss than without the flag fused=True. I will work on providing a minimal example but it could be a while.

**Comment by albanD** [comment]
cc @crcrpar who contributed the fused impl.

**Comment by ptrblck** [comment]
> The accuracy penalty only comes when both AMP & fused are simultaneously active. Individually they give identical losses but together the loss is much worse.

I cannot reproduce this claim and get the same results for the default and fused `Adam` using this simple code snippet:
```python
import torch
import torch.nn as nn

torch.manual_seed(2809)

# Creates model and optimizer in default precision
model = nn.Linear(10, 10).cuda()
optimizer = torch.optim.Adam(model.parameters(), lr=1., fused=False)

# Creates a GradScaler once at the beginning of training.
scaler = torch.cuda.amp.GradScaler()

data = torch.randn(10, 10, device='cuda')
target = torch.randn(10, 10, device='cuda')
loss_fn = nn.MSELoss()

epochs = 5
for epoch in range(epochs):
    optimizer.zero_grad()
    # Runs the forward pass with autocasting.
    with torch.cuda.amp.autocast(dtype=torch.float16):
        output = model(data)
        loss = loss_fn(output, target)

    # Scales loss.  Calls backward() on scaled loss to create scaled gradients.
    # Backward passes under autocast are not recommended.
    # Backward ops run in the same dtype autocast chose for corresponding forward ops.
    scaler.scale(loss).backward()

    # scaler.step() first unscales the gradients of the optimizer's assigned params.
    # If these gradients do not contain infs or NaNs, optimizer.step() is then called,
    # otherwise, optimizer.step() is skipped.
    scaler.step(optimizer)

    # Updates the scale for next iteration.
    scaler.update()
    
    print("epoch {}, loss {:.3f}".format(epoch, loss.item()))


# fused=False
# epoch 0, loss 1.302
# epoch 1, loss 11.289
# epoch 2, loss 2.388
# epoch 3, loss 2.994
# epoch 4, loss 5.163

# fused=True
# epoch 0, loss 1.302
# epoch 1, loss 11.289
# epoch 2, loss 2.388
# epoch 3, loss 2.994
# epoch 4, loss 5.163
```
in `1.13.0+cu117`.

Based on your `collect_env` output it also seems you are using multiple PyTorch installations:
```python
[pip3] torch==1.13.0
[pip3] torch-tb-profiler==0.4.0
[pip3] torchaudio==0.13.0
[pip3] torchmetrics==0.10.0
[pip3] torchvision==0.14.0
...
[conda] pytorch 1.13.0 py3.9_cuda11.7_cudnn8.5.0_0 pytorch
[conda] pytorch-cuda 11.7 h67b0de4_0 pytorch
[conda] pytorch-lightning 1.7.7 pypi_0 pypi
[conda] pytorch-mutex 1.0 cuda pytorch
[conda] torch 1.12.1+cu113 pypi_0 pypi
[conda] torch-tb-profiler 0.4.0 pypi_0 pypi
[conda] torchaudio 0.13.0 py39_cu117 pytorch
[conda] torchmetrics 0.10.0 pypi_0 pypi
[conda] torchvision 0.13.1+cu113 pypi_0 pypi
```
so I would also recommend to use one binary only.
